packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`unable to find project id: ${message}`);
+  }
+}
+
+async function findRootDir(): Promise<string> {
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    let dir = process.cwd();
+    while (dir !== path.dirname(dir)) {
+      const pkgPath = path.join(dir, '.vercel');
+      if (fs.existsSync(pkgPath)) {
+        return dir;
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new Error('token refresh only supported in node server environments');
+  }
+  throw new Error('unable to find root directory');
+}
+
+async function saveToken(
+  token: VercelTokenResponse,
+  projectId: string
+): Promise<void> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
+    );
+    const tokenJson = JSON.stringify(token);
+    const fs = await import('fs');
+    fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
+    fs.writeFileSync(tokenPath, tokenJson);
+    return;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to save token: ${message}`);
+  }
+}
+
+async function loadToken(
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const fs = await import('fs');
+
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
+    );
+    if (!fs.existsSync(tokenPath)) {
+      return null;
+    }
+    const tokenJson = fs.readFileSync(tokenPath, 'utf8');
+    const token = JSON.parse(tokenJson) as VercelTokenResponse;
+    return token;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to load token: ${message}`);
+  }
+}
+
+export function getTokenPayload(token: string): TokenPayload {
+  const tokenParts = token.split('.');
+  if (tokenParts.length !== 3) {
+    throw new Error('Invalid token');
+  }
+
+  const base64 = tokenParts[1].replace(/-/g, '+').replace(/_/g, '/');
+
+  if (isBrowserEnv()) {
+    if (typeof Uint8Array === undefined || typeof TextDecoder === undefined) {
+      throw new Error(
+        'The browser environment is not supported. Requires Uint8Array and TextDecoder'
+      );
+    }
+
+    const binaryString = window.atob(base64);
+    const toUtf8 = Uint8Array.from(binaryString, c => c.charCodeAt(0));
+    const decoder = new TextDecoder();
+    const payload = decoder.decode(toUtf8);
+    return JSON.parse(payload);
+  }
+
+  return JSON.parse(Buffer.from(base64, 'base64').toString('utf8'));
+}
+
+export function isExpired(token: TokenPayload): boolean {
+  return token.exp * 1000 < Date.now();
+}
```

Comment:
Instead of just checking that the token is currently expired, we could align the behavior here with what the CLI is doing: e.g., proactively refresh an OIDC token which will expire within the next 15 minutes. WDYT?

https://github.com/vercel/vercel/blob/main/packages/cli/src/util/env/refresh-oidc-token.ts#L10-L13